mri/reconstructors/self_calibrating.py

```diff
@@ -0,0 +1,179 @@
+        if linear_op.n_coils != 1:
+            raise ValueError("The value of n_coils for linear operation must "
+                             "be 1 for Self-Calibrating reconstruction!")
+        if gradient_method == 'analysis':
+            grad_class = GradSelfCalibrationAnalysis
+        elif gradient_method == 'synthesis':
+            grad_class = GradSelfCalibrationSynthesis
+        super(SelfCalibrationReconstructor, self).__init__(
+            fourier_op=fourier_op,
+            linear_op=linear_op,
+            prox_op=prox_op,
+            mu=mu,
+            gradient_method=gradient_method,
+            grad_class=grad_class,
+            lipschitz_cst=lipschitz_cst,
+            lips_calc_max_iter=lips_calc_max_iter,
+            num_check_lips=num_check_lips,
+            optimization_alg=optimization_alg,
+            init_gradient_op=False,
+            verbose=verbose,
+        )
+        self.smaps_gridding_method = smaps_gridding_method
+        self.smaps_extraction_mode = smaps_extraction_mode
+        if type(kspace_portion) == float:
+            self.kspace_portion = (kspace_portion,) * \
+                                  self.fourier_op.samples.shape[-1]
+        else:
+            self.kspace_portion = kspace_portion
+        if len(self.kspace_portion) != self.fourier_op.samples.shape[-1]:
+            raise ValueError("The k-space portion size used to estimate the" +
+                             "sensitivity information is not aligned with" +
+                             " the input dimension")
+        self.n_jobs = n_jobs
+
+    def reconstruct(self, kspace_data, x_init=None, num_iterations=100,
+                    reinit_grad_op=False, **kwargs):
+        """ This method calculates operator transform.
+        Parameters
+        ----------
+        kspace_data: np.ndarray
+            the acquired value in the Fourier domain.
+            this is y in above equation.
+        x_init: np.ndarray (optional, default None)
+            input initial guess image for reconstruction
+        num_iterations: int (optional, default 100)
+            number of iterations of algorithm
+        """
+        if self.fourier_op.n_coils != kspace_data.shape[0]:
+            raise ValueError("The provided number of coil (n_coils) do not "
+                             "match the data itself")
+        # Extract Sensitivity maps and initialize gradient
+        Smaps, _ = get_Smaps(
+            k_space=kspace_data,
+            img_shape=self.fourier_op.shape,
+            samples=self.fourier_op.samples,
+            thresh=self.kspace_portion,
+            min_samples=self.fourier_op.samples.min(axis=0),
+            max_samples=self.fourier_op.samples.max(axis=0),
+            mode=self.smaps_extraction_mode,
+            method=self.smaps_gridding_method,
+            n_cpu=self.n_jobs
+        )
+        self.extra_grad_args['Smaps'] = Smaps
+        self.initialize_gradient_op(**self.extra_grad_args)
```

Comment:
I have not added the method to lazily obtain Smaps and initialize or even give user the freedom to give the Smaps directly in this PR as that would make it more bulky